### ISSUE TITLE
Defining :run callbacks breaks whole callbacks infrastructure

### ISSUE THREAD

**Issue by katafrakt** [rails/rails#19405]
When you define `:run` callbacks, anything that calls any callbacks breaks with quite vague error message:

```
/home/katafrakt/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/activesupport-4.2.0/lib/active_support/core_ext/class/attribute.rb:106:in `block (2 levels) in class_attribute': wrong number of arguments (1 for 0) (ArgumentError)
	from /home/katafrakt/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/activesupport-4.2.0/lib/active_support/callbacks.rb:734:in `_run_run_callbacks'
	from /home/katafrakt/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/activesupport-4.2.0/lib/active_support/callbacks.rb:81:in `run_callbacks'
```

This behaviour can be easily reproduced with following script:

```ruby
gem 'activesupport', '4.2.0'
require 'active_support/all'

class Machine
	include ActiveSupport::Callbacks
	define_callbacks :run

	def start
		run_callbacks(:run) do
			puts "running"
		end
	end
end

m = Machine.new
m.start
```

This is caused by how `define callbacks` method works. In this case it creates class attribute `_run_callbacks` and defines method `_run_run_callbacks` that calls internal `_run_callbacks` method, which is now overriden by class attribute. In fact, running any callbacks (not only `run`) will result in error message above.

```ruby
      def define_callbacks(*names)
        options = names.extract_options!

        names.each do |name|
          class_attribute "_#{name}_callbacks"
          set_callbacks name, CallbackChain.new(name, options)

          module_eval <<-RUBY, __FILE__, __LINE__ + 1
            def _run_#{name}_callbacks(&block)
              _run_callbacks(_#{name}_callbacks, &block)
            end
          RUBY
        end
      end
```

It works with version 3.2.15, I'm not sure about intermediate versions. Master is also affected.

**Proposed solution**

Raise error with clear message when user tries to define `:run` callback. I can PR this unless someone comes up with better idea.

**Comment by tgxworld** [comment]
@katafrakt This was introduced in https://github.com/rails/rails/commit/796cab45561fce268aa74e6587cdb9cae3bb243e and I think it is an intended side effect.

**Comment by matthewd** [comment]
At a glance, it seems we can fix this by just renaming (the private) `_run_callbacks` method, so it doesn't conflict with the `"_#{name}_callbacks"` pattern.

**Comment by tgxworld** [comment]
@matthewd I reverted because I didn't see any benefit made from the previous optimization. On top of that, we're defining a more methods and making more methods calls everytime we run a callback.

**Comment by katafrakt** [comment]
I guess the fix won't end up in 4.2 line?

**Comment by rafaelfranca** [comment]
See the last comment on the pull request
